### PR TITLE
WIP: DataAccess module impovements

### PR DESCRIPTION
**Pull request by Fivell**
Goal of this PR is to provide flexible access to collection in different scopes of request livecycle.
DataAccess#find_collection  becomes more generic and can return different result depends on ```options``` argument.
Can be used in batch actions (no sorting,pagination needed), csv exports(no pagination needed), etc

### CODE REVIEW

**Comment by Fivell** [comment]
travis-ci failed  with cucumber 
> Failing Scenarios:
cucumber features/index/filters.feature:109 # Scenario: Clearing filter preserves custom parameters

trace

>ReferenceError: Can't find variable: options
ReferenceError: Can't find variable: options
    at http://127.0.0.1:56079/assets/active_admin.js:20289 in CheckboxToggler
    at http://127.0.0.1:56079/assets/active_admin.js:20779 in TableCheckboxToggler

**Comment by Fivell** [comment]
@timoschilling , can you help to understand why build failed ? 
I can't understand how batch actions can be related to form specs ...
>> ActiveAdmin::FormBuilder with has many inputs with allow destroy with an existing post should include a boolean field for _destroy

**Comment by timoschilling** [comment]
I think build fails while a dependency has changed. See #3781.
Problem is that I don't have the time to find the bug, at the moment.

**Comment by timoschilling** [comment]
please rebase this commit, #3781 is closed

**Comment by Fivell** [comment]
@timoschilling , feel like few comments and specs should be added, but I will have free time only next week, so let's wait a bit.

**Comment by Fivell** [comment]
@timoschilling , rebased

File: docs/9-batch-actions.md [hunk 1]
```diff
@@ -21,7 +21,7 @@ your desired batch action on all of them:
 ```ruby
 ActiveAdmin.register Post do
   batch_action :flag do |ids|
-    Post.find(ids).each do |post|
+    batch_action_collection.where(id: ids).each do |post|
```
Comment (timoschilling) [inline docs/9-batch-actions.md]: we should use `find` like here https://github.com/activeadmin/activeadmin/pull/3775/files#diff-16da2146f502eb423c49bb455a83c2f7R70

**Comment by timoschilling** [comment]
What are the open task?

**Comment by Fivell** [comment]
@timoschilling, we started this PR, having conversation in #3764 , also I hope it will help implement #2375  in future

File: docs/9-batch-actions.md [hunk 2]
```diff
@@ -21,7 +21,7 @@ your desired batch action on all of them:
 ```ruby
 ActiveAdmin.register Post do
   batch_action :flag do |ids|
-    Post.find(ids).each do |post|
+    batch_action_collection.find(id: ids).each do |post|
```
Comment (timoschilling) [inline docs/9-batch-actions.md]: You need to remove id:

**Comment by timoschilling** [comment]
#2375 could be a new PR. Is there something to do before merging this?

**Comment by Fivell** [comment]
@timoschilling , #2375  must be new PR, I don't feel like I can do this because of luck UI skills , hope somebody can help with it. 
before merging this we should decide if this collection of appliers is what we need for default batch actions collection https://github.com/Fivell/activeadmin/commit/d3b25b06318e5b256bf3f71d6b2ed69573896c4a#diff-95a1987f84cc2650e828b2d27c9a2f7fR29

Please take a look

**Comment by timoschilling** [comment]
Feedback:
1. Should we include `includes`? This can speed up the destroying of relations `has_many :orders, dependent: :destroy`
2. Should we add a `COLLECTION_APPLIES` constant to the `BatchController`? This will allow the customization.

Questions:
1. Do you want to write some more test?

**Comment by Fivell** [comment]
@timoschilling, I can't understand about ```includes```. In real life has_many relations are not needed for rendering , that's why there can't be such includes on resource level. 
 
We can also add ```COLLECTION_APPLIES``` for ```BatchController```, but it is also included as a part of  ```DataAccess```  module. 

We already can call 
 ```find_collection(only: [])``` if we don't wan't to use ```batch_colleciton```

**Comment by timoschilling** [comment]
`includes` prevent n+1 queries.

What I mean is:
```ruby
class ActiveAdmin::BatchActions::Controller
  COLLECTION_APPLIES = [:authorization_scope, :collection_decorator, :filtering, :scoping, :includes]

  def batch_action_collection
    find_collection(only: COLLECTION_APPLIES)
  end
end
```

You had many fights with the doc :wink:

**Comment by Fivell** [comment]
@timoschilling , I know that but I can't find real example when we should add ```include``` for batch_collection , when common scenario is changing some attribute of each element of collection.  I mean why we need make additional sql query for include if we don't care about it in all cases?

**Comment by timoschilling** [comment]
The most common Scenario:
Given:
```ruby
class Blog
 has_many :posts, dependent: :destroy
end

class Post
 has_many :comments, dependent: :destroy
end
```

DB:
2 Blogs with 2 Post with 2 Comments each

Exec batch delete of both with includes:
```sql
SELECT * FROM blogs WHERE id IN (1, 2);
SELECT * FROM posts WHERE blog_id IN (1, 2);
SELECT * FROM comments WHERE post_id IN (1, 2, 3, 4);
```

Exec batch delete of both without includes:
```sql
SELECT * FROM blogs WHERE id IN (1, 2);
SELECT * FROM posts WHERE blog_id = 1;
SELECT * FROM posts WHERE blog_id = 2;
SELECT * FROM comments WHERE post_id = 1;
SELECT * FROM comments WHERE post_id = 2;
SELECT * FROM comments WHERE post_id = 3;
SELECT * FROM comments WHERE post_id = 4;
```

Or a other example:
A Shop with Orders, each Order has n Items and a batch action "Send Invoice", here we can use a `includes(:items)`

**Comment by Fivell** [comment]
@timoschilling, lets look at example with blog/posts/comments 
and think how resource for blog will look like.

```ruby
ActiveAdmin.register Blog do 
  
    index do
       column :name
       column :description
       column :created_at 
    end
    
end
```
or something like this. What are saying is that for destroy optimisation I should add ```includes``` applier (but it doesn't exist at all on resource level)

ok lets add it and see what we have after

```ruby
ActiveAdmin.register Blog do 
   includes :posts
   
    index do
       column :name
       column :description
       column :created_at 
    end
    
end
```
ok so now we have faster batch destroy, but what about index ???  Index action uses includes applier also. We are loading posts every time on index page because index page uses ```find_collection``` without arguments.
 (there can be hundreds or thousands of posts) but why we are doing it ?  

Your example is not generic, to fix it we just need to override default destroy batch action, how it is already described in documentation, in other cases we bring optimisation to one place  , but remove it in another one.

**Comment by Fivell** [comment]
To clarify, ActiveAdmin DSL doesn't allow us to add ```includes``` applier for batch_action and not use it for simple index rendering

**Comment by timoschilling** [comment]
Ok you are right in the Blog example `includes` don't make sense for the index action.

But let us do this:
```ruby
class ActiveAdmin::BatchActions::Controller
  COLLECTION_APPLIES = [:authorization_scope, :collection_decorator, :filtering, :scoping, :includes]

  def batch_action_collection
    find_collection(only: COLLECTION_APPLIES)
  end
end
```

**Comment by Fivell** [comment]
@timoschilling , ok let's see what travis will tell

**Comment by Fivell** [comment]
@timoschilling , you're also hinting about docs,  how do you think I can improve it?

**Comment by timoschilling** [comment]
:+1: for
```ruby
      def batch_action_collection(only = COLLECTION_APPLIES)
        find_collection(only: only)
      end
```

**Comment by Fivell** [comment]
@timoschilling , what next?

**Comment by timoschilling** [comment]
Sorry I missed your last commit. I think we can merge or miss I something that you want to do?

**Comment by Fivell** [comment]
@timoschilling , I think it can be merged, or I should rebase and squash commits?

**Comment by timoschilling** [comment]
good work

**Pull request merged**